# Essential Linear Algebra for Finance

## 2.3     Applications to Lienar Portfolios  
  - 2.3.1  Covariance and Correlation Matrices  
  - 2.3.2  Portfolio Risk and Return in Matrix Notation  
  - 2.3.3  Positive Definiteness of Covariance and Correlation Matrices  
  - 2.3.4  Eigenvalues and Eigenvectors of Covariance and Correlation Matrices  

In [1]:
from visualisations import *

Measuring the market risk of a large portfolio requires mapping its risk factors. This process is depends on the type of assets in the portfolio (i.e. equity, bonds, options).  

This section will deal primarily with linear portfolios at the asset level. These portfolios can be represented as a linear combination (weighted sum) of the returns of the underlying assets such as cash, futures or forwards and are thus can be conveniently expressed in matrix notation. Portfolios with options and bonds are considered non-linear portfolios and capturing their risk factor sensitivities will be discussed in detail later on.

### 2.3.1 Covariance and Correlation Matrices

**Covariance Matrix**  

A symmetric matrix of variance and covariances of an $m \times 1$ vector of $m$ returns.  
The variances of the returns displayed in the diagonal and the covariances displayed in the other elements denoted as:  

$$
\mathbf{V} =
\begin{pmatrix}
\sigma_1^2 & \sigma_{12} & \cdots & \cdots & \sigma_{1m} \\
\sigma_{21} & \sigma_2^2 & \cdots & \cdots & \sigma_{2m} \\
\sigma_{31} & \sigma_{32} & \sigma_3^2 & \cdots & \sigma_{3m} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
\sigma_{m1} & \cdots & \cdots & \cdots & \sigma_m^2
\end{pmatrix}
$$

$\mathbf{V}$ is used to denote an arbitrary covariance matrix of returns particularly for returns of individual assets.  

$\Omega$ is reserved for covariance matrices of risk factor returns.

Any covariance matrix can be denoted by $\mathbf{V} = \mathbf{DCD}$ since  

$$
\begin{pmatrix}
\sigma_1^2 & \sigma_{12} & \cdots & \cdots & \sigma_{1m} \\
\sigma_{21} & \sigma_2^2 & \cdots & \cdots & \sigma_{2m} \\
\sigma_{31} & \sigma_{32} & \sigma_3^2 & \cdots & \sigma_{3m} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
\sigma_{m1} & \cdots & \cdots & \cdots & \sigma_m^2
\end{pmatrix}
=
\begin{pmatrix}
\sigma_1^2 & \varrho_{12} \sigma_1 \sigma_2 & \cdots & \cdots & \varrho_{1m} \sigma_1 \sigma_m \\
\varrho_{21} \sigma_2 \sigma_1 & \sigma_2^2 & \cdots & \cdots & \varrho_{2m} \sigma_2 \sigma_m \\
\varrho_{31} \sigma_3 \sigma_1 & \varrho_{32} \sigma_3 \sigma_2 & \sigma_3^2 & \cdots & \varrho_{3m} \sigma_3 \sigma_m\\
\vdots & \vdots & \vdots & \ddots & \vdots \\
\varrho_{m1} \sigma_m \sigma_1 & \cdots & \cdots & \cdots & \sigma_m^2
\end{pmatrix}
$$  

$\mathbf{D}$ a *diagonal matrix* of the standard deviations of the returns  
$\mathbf{C}$ a *correlation matrix* of the correlation of the returns  

$$
\begin{pmatrix}
\sigma_1^2 & \sigma_{12} & \cdots & \cdots & \sigma_{1m} \\
\sigma_{21} & \sigma_2^2 & \cdots & \cdots & \sigma_{2m} \\
\sigma_{31} & \sigma_{32} & \sigma_3^2 & \cdots & \sigma_{3m} \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
\sigma_{m1} & \cdots & \cdots & \cdots & \sigma_m^2
\end{pmatrix}
=
\begin{pmatrix}
\sigma_{1} & 0 & \cdots & \cdots & 0 \\
0 & \sigma_{2} & 0 & \cdots & 0 \\
0 & 0 & \sigma_{3} & \cdots & \cdots \\
\vdots & \vdots & \vdots & \ddots & 0 \\
0 & \cdots & \cdots & 0 & \sigma_{m}
\end{pmatrix}
\begin{pmatrix}
1 & \varrho_{12} & \cdots & \cdots & \varrho_{1m} \\
\varrho_{12} & 1 & \cdots & \cdots & \varrho_{2m} \\
\cdots & \cdots & \cdots & \cdots & \cdots  \\
\vdots & \vdots & \vdots & \ddots & \vdots \\
\varrho_{1m} & \varrho_{2m} & \cdots & \cdots & 1
\end{pmatrix}
\begin{pmatrix}
\sigma_{1} & 0 & \cdots & \cdots & 0 \\
0 & \sigma_{2} & 0 & \cdots & 0 \\
0 & 0 & \sigma_{3} & \cdots & \cdots \\
\vdots & \vdots & \vdots & \ddots & 0 \\
0 & \cdots & \cdots & 0 & \sigma_{m}
\end{pmatrix}
$$

Find the annual covariance matrix  

| Asset       | Volatility       | Assets            |  Correlation |
|:------------|-----------------:|:-----------------:|-------------:|
|Asset 1      |20%               |Asset 1 - asset 2  | 0.8          |
|Asset 2      |10%               |Asset 1 - asset 3  | 0.5          |
|Asset 3      |15%               |Asset 3 - asset 2  | 0.3          |  

$\mathbf{D} = \begin{bmatrix} 0.2 & 0 & 0 \\ 0 & 0.1 & 0 \\ 0 & 0 & 0.15 \end{bmatrix}$   $\mathbf{C} = \begin{bmatrix} 1 & 0.8 & 0.5 \\ 0.8 & 1 & 0.3 \\ 0.5 & 0.3 & 1 \end{bmatrix}$ 

In [2]:
# Import numpy 
import numpy as np

# Create diagonal matrix of portfolio variances
variance = np.array([[0.2, 0, 0],
                    [0, 0.1, 0],
                    [0, 0, 0.15]])

# Create correlation matrix
correlation = np.array([[1, 0.8, 0.5],
                        [0.8, 1, 0.3],
                        [0.5, 0.3, 1]])

# Find the annual covariance matrix DCD
covariance = variance @ correlation @ variance

# Print results
print(covariance)

[[0.04   0.016  0.015 ]
 [0.016  0.01   0.0045]
 [0.015  0.0045 0.0225]]


Using python's numpy library to find the annual covariance matrix $\mathbf{DCD}$ to find 

$$
\begin{bmatrix} 0.2 & 0 & 0 \\ 0 & 0.1 & 0 \\ 0 & 0 & 0.15 \end{bmatrix}
\begin{bmatrix} 1 & 0.8 & 0.5 \\ 0.8 & 1 & 0.3 \\ 0.5 & 0.3 & 1 \end{bmatrix}
\begin{bmatrix} 0.2 & 0 & 0 \\ 0 & 0.1 & 0 \\ 0 & 0 & 0.15 \end{bmatrix}
=
\begin{bmatrix} 0.04 & 0.016 & 0.015 \\ 0.016 & 0.01 &  0.0045 \\ 0.015 & 0.0045 & 0.0225 \end{bmatrix}

### 2.3.2 Portfolio Risk and Return in Matrix Notation

In 1.3 we derived the percentage return $R$ for a linear portfolio as the weighted sum of the percentage returns for each asset.  

$$R = \sum_{i = 1}^{k}w_{i}R_{i}$$ 

We can write this in matrix form as  

$$R = \mathbf{w}^\top \mathbf{r}$$
$\mathbf{w} = \begin{bmatrix} w_{1} , & \cdots , & w_{k} \end{bmatrix}^\top$ is the $k \times 1$ vector of portfolio weights  

$\mathbf{r} = \begin{bmatrix} R_{1} , & \cdots & , R_{k} \end{bmatrix}^\top$ is the $k \times 1$ vector of asset returns

The following expression represents the variance of the portfolio returns in sigma notation similar to the return example from 1.3

$$
V(R) = \sum_{i=1}^{k} w_i^2 V(R_i) + \sum_{i=1}^{k} \sum_{j=1}^{k} w_i w_j \, \text{Cov}(R_i, R_j).
$$  

It is much more convenient to write in matrix form with the portfolio variance as a quadratic form  

$$
V(R) = \mathbf{w}^\top \mathbf{V} \mathbf{w}
$$  

with $\mathbf{w}$ representing the vector of portfolio weights and $\mathbf{V}$ as the covariance matrix of asset returns.

$\mathbf{V} = \mathbf{DCD}$ provides another way to denote portfolio variance

$V(R) = \mathbf{x}^\top \mathbf{Cx}$,  

$\mathbf{x} = \mathbf{Dw}$  

The variance of a portfolio's returns can be expressed as  

- The quadratic form of the covariance matrix $\mathbf{V}$ and vector of portfolio weights $\mathbf{w}$
- The quadratic form of the correlation matrix $\mathbf{C}$ and the vector $\mathbf{x}$ where the weight of each asset multiplied by the standard deviation of that asset's return.

**Volatility of Returns and Volatility of P&L**  

Volatility refers to the annualized standard deviation or square root of the variance.  

Assuming constant weights the volatility of the portfolio P&L is found by multiplying the volatility of the returns by the current value of the portfolio.

|Assets | Volatility | Correlation | Value | Weight |
|:------|-----------:|:-----:|:-----:|:------:|
|Asset 1 | 10 %| 0.5 |    €2,000,000 | 40%|
|Asset 2 | 20 %| 0.5 |€3,000,000 |  60% |

$\mathbf{V} = \mathbf{DCD} = \begin{bmatrix} 0.10 & 0 \\ 0 & 0.20 \end{bmatrix} \begin{bmatrix} 1 & 0.5 \\ 0.5 & 1 \end{bmatrix} \begin{bmatrix} 0.10 & 0 \\ 0 & 0.20 \end{bmatrix} = \begin{bmatrix} 0.01 & 0.01 \\ 0.01 & 0.04 \end{bmatrix} = \begin{bmatrix} 1 & 1 \\ 1 & 4 \end{bmatrix} \times 10^{-2}$  

And the portfolio return  

$\mathbf{w}^\top \mathbf{Vw} = \begin{bmatrix} 0.4 & 0.6 \end{bmatrix} \begin{bmatrix} 1 & 1 \\ 1 & 4 \end{bmatrix} \begin{bmatrix} 0.4 \\ 0.6 \end{bmatrix} \times 10^{-2} = 2.08 \times 10^{-2}$  

or equivalently  

$\mathbf{x} = \mathbf{Dw} = \begin{bmatrix} 0.10 & 0 \\ 0 & 0.20 \end{bmatrix} \begin{bmatrix} 0.4 \\ 0.6 \end{bmatrix} = \begin{bmatrix} 0.04 \\ 0.12 \end{bmatrix}$

$\mathbf{x}^\top \mathbf{Cx} = \begin{bmatrix} 0.04 & 0.12 \end{bmatrix} \begin{bmatrix} 1 & 0.5 \\ 0.5 & 1 \end{bmatrix} \begin{bmatrix} 0.04 \\ 0.12 \end{bmatrix} = 0.0208$  

Both methods give the portfolio return's variance of $\sigma^{2}$ = 0.0208 and volatility $\sigma = 14.42 \%$.  

The portfolio's P&L has volatility $14.42 \% \times €5,000,000 = €721,000$


### 2.3.3 Positive Definiteness of Covariance and Correlation Matrices